File: backend/app/schemas/parcels.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ParcelSource = Literal["upload", "drawn"]
MAX_FEATURES_PER_REQUEST = 500
MAX_VERTICES_PER_PARCEL = 5000
MAX_UPLOAD_BYTES = 5 * 1024 * 1024


class ParcelProps(BaseModel):
    id: uuid.UUID
    name: str
    category: str
    notes: str | None
    area_m2: float
    source: str
    source_ref: str | None
    created_at: datetime
    updated_at: datetime


class ParcelFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    id: uuid.UUID
    geometry: dict[str, Any]
    properties: ParcelProps


class ParcelFeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[ParcelFeature]
    total: int
    page: int
    page_size: int


class ParcelCreateFeature(BaseModel):
    """One incoming feature. `name`/`category` come from properties or the request defaults."""

    type: Literal["Feature"] = "Feature"
    geometry: dict[str, Any]
    properties: dict[str, Any] = Field(default_factory=dict)


class ParcelCreateRequest(BaseModel):
    """Body of POST /parcels: a GeoJSON FeatureCollection (uploaded or drawn) plus defaults.

    Per-feature `properties.name` / `properties.category` / `properties.notes` win over the
    request-level defaults. Invalid features are skipped and reported (appflow Flow B step 5).
    """

    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[ParcelCreateFeature] = Field(min_length=1, max_length=MAX_FEATURES_PER_REQUEST)
    source: ParcelSource = "upload"
    source_ref: str | None = Field(default=None, max_length=200)
    name: str | None = Field(default=None, min_length=1, max_length=200)
    category: str | None = Field(default=None, min_length=1, max_length=100)
    notes: str | None = Field(default=None, max_length=2000)


class ParcelSkipped(BaseModel):
    index: int
    name: str | None
    reason: str


class ParcelCreateResult(BaseModel):
    created: list[ParcelFeature]
    skipped: list[ParcelSkipped]


class ParcelPatch(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    category: str | None = Field(default=None, min_length=1, max_length=100)
    notes: str | None = Field(default=None, max_length=2000)
    # Geometry redraw is only allowed for drawn parcels (appflow Flow B step 9).
    geometry: dict[str, Any] | None = None


class ParcelDeleteResult(BaseModel):
    deleted: uuid.UUID
    cascade: bool
    scans_removed: int
    detections_removed: int

File: frontend/src/api/parcels.ts
import type { Geometry } from 'geojson';
import { apiFetch } from './client';
import type { ParcelCollection, ParcelProps } from './detections';

export type { ParcelCollection, ParcelProps };

export interface ParcelFeature {
  type: 'Feature';
  id: string;
  geometry: Geometry;
  properties: ParcelProps;
}

export interface ParcelCreateFeature {
  type: 'Feature';
  geometry: Geometry;
  properties: { name?: string; category?: string; notes?: string };
}

export interface ParcelCreateResult {
  created: ParcelFeature[];
  skipped: { index: number; name: string | null; reason: string }[];
}

export interface ParcelDeleteResult {
  deleted: string;
  cascade: boolean;
  scans_removed: number;
  detections_removed: number;
}

export const CATEGORIES = ['wetland', 'forest', 'reserve', 'lake', 'coastal', 'other'] as const;

export const listParcels = (q = '') =>
  apiFetch<ParcelCollection>(`parcels?page_size=200${q ? `&q=${encodeURIComponent(q)}` : ''}`);
export const getParcel = (id: string) => apiFetch<ParcelFeature>(`parcels/${id}`);
export const createParcels = (
  features: ParcelCreateFeature[],
  source: 'upload' | 'drawn',
  defaults: { name?: string; category?: string; source_ref?: string } = {},
) =>
  apiFetch<ParcelCreateResult>('parcels', {
    method: 'POST',
    body: JSON.stringify({ type: 'FeatureCollection', features, source, ...defaults }),
  });
export const patchParcel = (
  id: string,
  body: { name?: string; category?: string; notes?: string | null; geometry?: Geometry },
) => apiFetch<ParcelFeature>(`parcels/${id}`, { method: 'PATCH', body: JSON.stringify(body) });
export const deleteParcel = (id: string, cascade = false) =>
  apiFetch<ParcelDeleteResult>(`parcels/${id}${cascade ? '?cascade=true' : ''}`, {
    method: 'DELETE',
  });

// Phase 9.3 — per-parcel monthly timeline
export interface TimelineMonth {
  month: string; // YYYY-MM-01
  built_frac: number | null;
  ndvi_mean: number | null;
  valid_frac: number | null;
  n_scenes: number;
}
export interface TimelineJob {
  status: 'running' | 'done' | 'failed';
  progress: number;
  message: string | null;
  months_total: number;
  months_done: number;
  started_at: string;
  finished_at: string | null;
}
export interface Timeline {
  parcel_id: string;
  t_bui: number;
  months: TimelineMonth[];
  onset_month: string | null;
  job: TimelineJob | null;
  running: boolean;
  note: string;
}
export const getTimeline = (id: string) => apiFetch<Timeline>(`parcels/${id}/timeline`);
export const startTimeline = (id: string, months = 36) =>
  apiFetch<Timeline>(`parcels/${id}/timeline`, {
    method: 'POST',
    body: JSON.stringify({ months }),
  });
